cli/src/cli_util.rs

```diff
@@ -3062,6 +3063,102 @@ impl fmt::Display for RemoteBookmarkNamePattern {
     }
 }
 
+/// Computes the location (new parents and new children) to place commits.
+///
+/// The `destination` argument is mutually exclusive to the `insert_after` and
+/// `insert_before` arguments.
+pub fn compute_commit_location(
+    ui: &mut Ui,
+    workspace_command: &WorkspaceCommandHelper,
+    destination: Option<&[RevisionArg]>,
+    insert_after: Option<&[RevisionArg]>,
+    insert_before: Option<&[RevisionArg]>,
+    commit_description: &str,
+) -> Result<(Vec<Commit>, Vec<Commit>), CommandError> {
+    let resolve_revisions =
+        |revisions: Option<&[RevisionArg]>| -> Result<Option<Vec<Commit>>, CommandError> {
+            if let Some(revisions) = revisions {
+                Ok(Some(
+                    workspace_command
+                        .resolve_some_revsets_default_single(ui, revisions)?
+                        .into_iter()
+                        .collect_vec(),
+                ))
+            } else {
+                Ok(None)
+            }
+        };
+    let destination_commits = resolve_revisions(destination)?;
+    let after_commits = resolve_revisions(insert_after)?;
+    let before_commits = resolve_revisions(insert_before)?;
+
+    let (new_parents, new_children) = match (destination_commits, after_commits, before_commits) {
+        (Some(destination_commits), None, None) => (destination_commits, vec![]),
+        (None, Some(after_commits), Some(before_commits)) => (after_commits, before_commits),
+        (None, Some(after_commits), None) => {
+            let new_children: Vec<_> =
+                RevsetExpression::commits(after_commits.iter().ids().cloned().collect_vec())
+                    .children()
+                    .evaluate(workspace_command.repo().as_ref())?
+                    .iter()
+                    .commits(workspace_command.repo().store())
+                    .try_collect()?;
+
+            (after_commits, new_children)
+        }
+        (None, None, Some(before_commits)) => {
+            // Not using `RevsetExpression::parents` here to persist the order of parents
+            // specified in `before_commits`.
+            let new_parent_ids = before_commits
+                .iter()
+                .flat_map(|commit| commit.parent_ids().iter().cloned())
+                .unique();
+            let new_parents: Vec<_> = new_parent_ids
+                .map(|commit_id| workspace_command.repo().store().get_commit(&commit_id))
+                .try_collect()?;
+
+            (new_parents, before_commits)
+        }
+        _ => unreachable!(),
```

Comment:
I think we should return an error rather than panicking here. It seems like there are two scenarios:

1. The caller didn't provide any of the dest, before, or after arguments.
2. The caller provided dest and one or more of the other two.

Can we return an Internal CommandError with a useful message in those cases?

Reply:
I didn't do this initially because it was expected that the arguments would be passed in directly from the clap command, and clap would have validated these conditions. I'm not sure we need to create a `CommandError` since reaching this case would be an internal error, and probably doesn't have a useful message for external users?

Reply:
Users should never see it, but implementers might if they make a mistake (i.e. pass the default destination arg like you called out below).

You can treat it as an optional nit if you want, I just think it's better not to panic if it can be avoided.

Reply:
I would use `panic!()` or `unreachable!()` with message in that case.